datalad/support/tests/test_annexrepo.py

```diff
@@ -2235,3 +2235,26 @@ def test_ro_operations(path):
 
     # just check that all is good again
     repo2.repo_info()
+
+
+@with_tempfile
+def test_annex_cmd_expect_stderr(path):
+    # test, that log message about stderr is logged on debug level if expect_stderr was set to True,
+    # warning level else
+
+    from re import DOTALL
+
+    repo = AnnexRepo(path)
+
+    with swallow_logs(logging.DEBUG) as cml:
+        repo._run_annex_command_json('add', ['non-existing'], expect_stderr=True)
+        # message shows up at DEBUG level:
+        assert_re_in(r".*\[DEBUG\][^[]*git-annex: add: 1 failed", cml.out, flags=DOTALL)
```

Comment:
Fine as is, but the helper cml.assert_logged() could also be used here, which I find slightly more readable but always forget about myself.

Reply:
It would be nice if it could be used, but it can't ATM. Requires a rewrite of assert_logged, which isn't flexible enough for that. It will prepend the message I'm looking for with a fixed prefix of for the level. But this prefix doesn't account for other things in the captured log before that. It would assume the captured log actually starts with `[DEBUG]` for example. But that doesn't work if there are other things logged at higher level before. In short: I tried, but gave up on it, since I was already spending more time than I wanted. Only tackled this one, because I thought: "Well, one-liner. Just spend those 5 minutes." Always turns out to be wrong ;-)

Reply:
> It would be nice if it could be used, but it can't ATM. Requires a rewrite of assert_logged, which isn't flexible enough for that. It will prepend the message I'm looking for with a fixed prefix of for the level. But this prefix doesn't account for other things in the captured log before that. It would assume the captured log actually starts with `[DEBUG]` for example. But that doesn't work if there are other things logged at higher level before.

`match=False` should cover that:

```diff
diff --git a/datalad/support/tests/test_annexrepo.py b/datalad/support/tests/test_annexrepo.py
index ddacc5ef8..99bd2b10a 100644
--- a/datalad/support/tests/test_annexrepo.py
+++ b/datalad/support/tests/test_annexrepo.py
@@ -2249,8 +2249,10 @@ def test_annex_cmd_expect_fail(path):
     with swallow_logs(logging.DEBUG) as cml:
         repo._run_annex_command_json('add', ['non-existing'], expect_fail=True)
         # message shows up at DEBUG level:
-        assert_re_in(r".*\[DEBUG\][^[]*git-annex: add: 1 failed", cml.out, flags=DOTALL)
+        cml.assert_logged(r"[^[]*git-annex: add: 1 failed",
+                          level="DEBUG", match=False)
     with swallow_logs(logging.DEBUG) as cml:
         repo._run_annex_command_json('add', ['non-existing'], expect_fail=False)
         # message shows up at WARNING level
-        assert_re_in(r".*\[WARNING\][^[]*git-annex: add: 1 failed", cml.out, flags=DOTALL)
+        cml.assert_logged(r"[^[]*git-annex: add: 1 failed",
+                          level="WARNING", match=False)
```

> In short: I tried, but gave up on it, since I was already spending more time than I wanted. Only tackled this one, because I thought: "Well, one-liner. Just spend those 5 minutes." Always turns out to be wrong ;-)

A familiar feeling :)

Reply:
Well, that seems true. Still don't like assert_logged much, since it fails to properly check a log entry at a time (which is why `r"[^[]*` is there), but I'll honor you pointing that out by using it :-)